backend/src/tx.rs

```diff
@@ -0,0 +1,136 @@
+use subxt::ext::codec::Encode;
+use subxt::ext::sp_runtime::traits::{IdentifyAccount, Verify};
+use subxt::ext::{sp_core, sp_runtime};
+use subxt::{config::polkadot::PolkadotExtrinsicParams, config::Config};
+
+use crate::configuration::Configuration;
+use crate::error::AppError;
+use crate::utils::{calculate_signature, get_current_block, get_tx_counter};
+
+use self::kilt::runtime_types;
+use self::kilt::runtime_types::did::did_details::DidAuthorizedCallOperation;
+
+#[subxt::subxt(runtime_metadata_path = "./metadata.scale")]
+pub mod kilt {}
+
+pub type ProxyType = kilt::runtime_types::peregrine_runtime::ProxyType;
+pub type RuntimeCall = kilt::runtime_types::peregrine_runtime::RuntimeCall;
+pub type RuntimeEvent = kilt::runtime_types::peregrine_runtime::RuntimeEvent;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct KiltConfig;
+impl Config for KiltConfig {
+    type Hash = sp_core::H256;
+    type Hasher = <subxt::config::SubstrateConfig as Config>::Hasher;
+    type AccountId = <<Self::Signature as Verify>::Signer as IdentifyAccount>::AccountId;
+    type Address = sp_runtime::MultiAddress<Self::AccountId, ()>;
+    type Header = subxt::config::substrate::SubstrateHeader<u64, Self::Hasher>;
+    type Signature = sp_runtime::MultiSignature;
+    type ExtrinsicParams = PolkadotExtrinsicParams<Self>;
+}
+
+pub async fn create_claim(
+    claim_hash: sp_core::H256,
+    ctype_hash: sp_core::H256,
+    config: Configuration,
+) -> Result<Vec<u8>, AppError> {
+    let api = config.get_client().await?;
+    let payer = config.get_payer_signer()?;
+    let payer_account = payer.account_id();
+    let tx_counter = get_tx_counter(&config).await?;
+    let block_number = get_current_block(&config).await?;
+
+    let call = RuntimeCall::Attestation(runtime_types::attestation::pallet::Call::add {
+        claim_hash,
+        ctype_hash,
+        authorization: None,
+    });
+
+    let did_call = DidAuthorizedCallOperation {
+        did: config.get_did()?,
+        tx_counter,
+        block_number,
+        call,
+        submitter: payer_account.clone().into(),
+    };
+
+    let encoded_call = did_call.encode();
+
+    let signature = calculate_signature(&did_call.encode(), config)?;
+    let final_tx = kilt::tx().did().submit_did_call(did_call, signature);
+    let events = api
+        .tx()
+        .sign_and_submit_then_watch_default(&final_tx, &payer)
+        .await?
+        .wait_for_finalized_success()
```

Comment:
this is too long to have it hanging in the request until it is finished. There is so much that can happen in those 20 seconds. IMO tx submission and status reporting needs to be done asyncronously. I.e. the attestation request needs to have a field `tx_state` which is either `in_flight`, `in_block`, `succeeded` or `failed`.

Reply:
added an enum which represents the tx state: [fa26a9c](https://github.com/KILTprotocol/attester/pull/1/commits/fa26a9c7aa71e8b501a4e6269db6c02e1ec19719)

Reply:
custom enums are not really great supported in sql. I have to change the return statement in the querys to: [3c49e88](https://github.com/KILTprotocol/attester/pull/1/commits/3c49e88c7d616151b9639646a1db0415ecbb6d19)

it is an open issue in sqlx: https://github.com/launchbadge/sqlx/issues/1004